tools/src/main/scala/scala/scalanative/optimizer/pass/SRemOverflow.scala

```diff
@@ -0,0 +1,68 @@
+package scala.scalanative.optimizer.pass
+
+import scala.scalanative.nir
+import scala.scalanative.nir._
+import scala.scalanative.optimizer.analysis.ClassHierarchy.Top
+import scala.scalanative.optimizer.{Pass, PassCompanion}
+import scala.scalanative.tools.Config
+
+/**
+ * Detects taking remainder for division by -1 and replaces it by division by 1 which can't overflow.
+ *
+ *
+ * We implement '%' (remainder) with LLVM's 'srem' and it can overflow for cases:
+ * Int.MinValue % -1
+ * Long.MinValue % -1
+ * E.g. On x86_64 'srem' might get translated to 'idiv' which computes both quotient and remainder at once
+ * and quotient can overflow.
+ */
+class SRemOverflow(implicit top: Top) extends Pass {
+
+  override def onInsts(insts: Seq[Inst]): Seq[Inst] = {
+    val buf = new nir.Buffer
+    import buf._
+
+    def matchSrem(IntType: Type,
+                  makeVal: Int => Val): PartialFunction[Inst, Unit] = {
+
+      case Inst.Let(name, sremBin @ Op.Bin(Bin.Srem, IntType, _, divisor)) =>
```

Comment:
Instead of taking `IntType` as an argument, you can match using `case ... Type.I(width) ... if width == 32 || width == 64`. The whole helper function can then be hand inlined into foreach.

Reply:
Sure, done.

I'm still always constructing `Int.Val` and I was worried it wouldn't work for longs, but turns out it doesn't seem to matter.